File: templates/problems/07_trees/02_max_depth/go/solution.go
package max_depth

// solveMaxDepth returns the maximum depth of a binary tree.
// Time: O(n), Space: O(h) where h is the height of the tree.
func solveMaxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	left := solveMaxDepth(root.Left)
	right := solveMaxDepth(root.Right)
	if left > right {
		return left + 1
	}
	return right + 1
}

File: templates/problems/07_trees/02_max_depth/python/solution.py
from helpers import TreeNode, Optional


def solve_max_depth(root: Optional[TreeNode]) -> int:
    """Return the maximum depth of a binary tree.

    Time: O(n), Space: O(h) where h is the height of the tree.
    """
    if root is None:
        return 0
    left = solve_max_depth(root.left)
    right = solve_max_depth(root.right)
    return left + 1 if left > right else right + 1
